lms/djangoapps/courseware/tests/test_date_summary.py

```diff
@@ -650,6 +670,68 @@ def test_date_with_existing_schedule(self):
         block = VerifiedUpgradeDeadlineDate(course, enrollment.user)
         self.assertEqual(block.date, expected)
 
+    @ddt.data(
+        # (org enabled, org opt-out, course enabled, course opt-out, expected dynamic deadline)
+        (False, False, False, False, False, True),
+        (False, False, False, False, True, True),
+        (False, False, False, True, False, True),
+        (False, False, False, True, True, False),
+        (False, False, True, False, False, True),
+        (False, False, True, False, True, True),
+        (False, False, True, True, False, True),
+        (False, False, True, True, True, False),
+        (False, True, False, False, False, True),
+        (False, True, False, False, True, True),
+        (False, True, False, True, False, True),
+        (False, True, False, True, True, False),  # course-level overrides org-level
+        (False, True, True, False, False, False),
+        (False, True, True, False, True, False),
+        (False, True, True, True, False, True),  # course-level overrides org-level
+        (False, True, True, True, True, False),
+
+        (True, False, False, False, False, True),
+        (True, False, False, False, True, True),
+        (True, False, False, True, False, True),
+        (True, False, False, True, True, False),
+        (True, False, True, False, False, True),
+        (True, False, True, False, True, True),
+        (True, False, True, True, False, True),
+        (True, False, True, True, True, False),
+        (True, True, False, False, False, True),
+        (True, True, False, False, True, True),
+        (True, True, False, True, False, True),
+        (True, True, False, True, True, False),  # course-level overrides org-level
+        (True, True, True, False, False, False),
+        (True, True, True, False, True, False),
+        (True, True, True, True, False, True),  # course-level overrides org-level
+        (True, True, True, True, True, False),
```

Comment:
There's got to be a better way to enumerate all of these, anyone know a good way?

Reply:
`itertools.product` maybe?

Reply:
Unfortunately, the last boolean in the list depends on the combination of the others. This is as concise as I could get it, but I'm not really sure it's anymore readable than just being explicit with the full list...

```python
[(case + (False,)
 if (case[3] and case[4]) or (case[1] and case[2] and not (case[3] and not case[4]))
 else case + (True,))
 for case in itertools.product([False, True], repeat=5)]
```

Or, in function form:

```python
def get_test_cases(self):
    ret_val = list()
    product = itertools.product([False, True], repeat=5)
    for case in product:
        if (case[3] and case[4]) or (case[1] and case[2] and not (case[3] and not case[4])):
            ret_val.append(case + (False,))
        else:
            ret_val.append(case + (True,))
    return ret_val
```

Reply:
Ah, right. In that case, I think the explicit list is clearer. If you collapsed down the flags to helpers on the config objects, you might be able to reduce the number of args to iterate through, but I don't think that's necessary to merge.

Reply:
(It might be necessary when we get one or two more config classes to check. :-P)